File: RxCamera/src/main/api14/com/vondear/camera/TextureViewPreview.java
package com.vondear.camera;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Matrix;
import android.graphics.SurfaceTexture;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;

@TargetApi(14)
class TextureViewPreview extends PreviewImpl {

    private final TextureView mTextureView;

    private int mDisplayOrientation;

    TextureViewPreview(Context context, ViewGroup parent) {
        final View view = View.inflate(context, R.layout.texture_view, parent);
        mTextureView = (TextureView) view.findViewById(R.id.texture_view);
        mTextureView.setSurfaceTextureListener(new TextureView.SurfaceTextureListener() {

            @Override
            public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
                setSize(width, height);
                configureTransform();
                dispatchSurfaceChanged();
            }

            @Override
            public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
                setSize(width, height);
                configureTransform();
                dispatchSurfaceChanged();
            }

            @Override
            public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
                setSize(0, 0);
                return true;
            }

            @Override
            public void onSurfaceTextureUpdated(SurfaceTexture surface) {
            }
        });
    }

    // This method is called only from Camera2.
    @TargetApi(15)
    @Override
    void setBufferSize(int width, int height) {
        mTextureView.getSurfaceTexture().setDefaultBufferSize(width, height);
    }

    @Override
    Surface getSurface() {
        return new Surface(mTextureView.getSurfaceTexture());
    }

    @Override
    SurfaceTexture getSurfaceTexture() {
        return mTextureView.getSurfaceTexture();
    }

    @Override
    View getView() {
        return mTextureView;
    }

    @Override
    Class getOutputClass() {
        return SurfaceTexture.class;
    }

    @Override
    void setDisplayOrientation(int displayOrientation) {
        mDisplayOrientation = displayOrientation;
        configureTransform();
    }

    @Override
    boolean isReady() {
        return mTextureView.getSurfaceTexture() != null;
    }

    /**
     * Configures the transform matrix for TextureView based on {@link #mDisplayOrientation} and
     * the surface size.
     */
    void configureTransform() {
        Matrix matrix = new Matrix();
        if (mDisplayOrientation % 180 == 90) {
            final int width = getWidth();
            final int height = getHeight();
            // Rotate the camera preview when the screen is landscape.
            matrix.setPolyToPoly(
                    new float[]{
                            0.f, 0.f, // top left
                            width, 0.f, // top right
                            0.f, height, // bottom left
                            width, height, // bottom right
                    }, 0,
                    mDisplayOrientation == 90 ?
                            // Clockwise
                            new float[]{
                                    0.f, height, // top left
                                    0.f, 0.f, // top right
                                    width, height, // bottom left
                                    width, 0.f, // bottom right
                            } : // mDisplayOrientation == 270
                            // Counter-clockwise
                            new float[]{
                                    width, 0.f, // top left
                                    width, height, // top right
                                    0.f, 0.f, // bottom left
                                    0.f, height, // bottom right
                            }, 0,
                    4);
        } else if (mDisplayOrientation == 180) {
            matrix.postRotate(180, getWidth() / 2, getHeight() / 2);
        }
        mTextureView.setTransform(matrix);
    }

}

File: RxCamera/src/main/api14/com/tamsiree/camera/TextureViewPreview.kt
package com.tamsiree.camera

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Matrix
import android.graphics.SurfaceTexture
import android.view.Surface
import android.view.TextureView
import android.view.TextureView.SurfaceTextureListener
import android.view.View
import android.view.ViewGroup

@TargetApi(14)
internal class TextureViewPreview(context: Context?, parent: ViewGroup?) : PreviewImpl() {
    private val mTextureView: TextureView
    private var mDisplayOrientation = 0

    // This method is called only from Camera2.
    @TargetApi(15)
    public override fun setBufferSize(width: Int, height: Int) {
        mTextureView.surfaceTexture.setDefaultBufferSize(width, height)
    }

    public override fun getSurface(): Surface {
        return Surface(mTextureView.surfaceTexture)
    }

    public override fun getSurfaceTexture(): SurfaceTexture {
        return mTextureView.surfaceTexture
    }

    public override fun getView(): View {
        return mTextureView
    }

    public override fun getOutputClass(): Class<*> {
        return SurfaceTexture::class.java
    }

    public override fun setDisplayOrientation(displayOrientation: Int) {
        mDisplayOrientation = displayOrientation
        configureTransform()
    }

    public override fun isReady(): Boolean {
        return mTextureView.surfaceTexture != null
    }

    /**
     * Configures the transform matrix for TextureView based on [.mDisplayOrientation] and
     * the surface size.
     */
    fun configureTransform() {
        val matrix = Matrix()
        if (mDisplayOrientation % 180 == 90) {
            val width = width
            val height = height
            // Rotate the camera preview when the screen is landscape.
            matrix.setPolyToPoly(floatArrayOf(
                    0f, 0f,  // top left
                    width.toFloat(), 0f,  // top right
                    0f, height.toFloat(),  // bottom left
                    width.toFloat(), height.toFloat()), 0,
                    if (mDisplayOrientation == 90) floatArrayOf(
                            0f, height.toFloat(),  // top left
                            0f, 0f,  // top right
                            width.toFloat(), height.toFloat(),  // bottom left
                            width.toFloat(), 0f) else floatArrayOf(
                            width.toFloat(), 0f,  // top left
                            width.toFloat(), height.toFloat(),  // top right
                            0f, 0f,  // bottom left
                            0f, height.toFloat()), 0,
                    4)
        } else if (mDisplayOrientation == 180) {
            matrix.postRotate(180f, width / 2.toFloat(), height / 2.toFloat())
        }
        mTextureView.setTransform(matrix)
    }

    init {
        val view = View.inflate(context, R.layout.texture_view, parent)
        mTextureView = view.findViewById<View>(R.id.texture_view) as TextureView
        mTextureView.surfaceTextureListener = object : SurfaceTextureListener {
            override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
                setSize(width, height)
                configureTransform()
                dispatchSurfaceChanged()
            }

            override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
                setSize(width, height)
                configureTransform()
                dispatchSurfaceChanged()
            }

            override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
                setSize(0, 0)
                return true
            }

            override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {}
        }
    }
}